### ISSUE TITLE
add a rule to warn/error on usage of for(.. in)

### ISSUE THREAD

**Issue by jhford** [eslint/eslint#3060]
##### Which version of ESLint you're currently using
```
$ eslint -v
v0.17.1
```
##### If you have any interest in addressing the issue yourself
I have already written the code and some docs and posted it in the now closed https://github.com/eslint/eslint/pull/3044

##### The problem you want to solve.
The `for (x in y) {}` construct is nuanced.  I'm not an expert js developer, but my understanding is that it is not really all that great to use.  With ES6 becoming more common, the `for (x of y) {}` construct is becoming more widely accepted.  Given that for-of is to JS what for-in is for most other languages, it's really easy to accidentally use for-in when you meant for-of.  For people, like myself, who work in more than one language having this confusion is very annoying.

I wrote this rule after causing a bug in one of my projects:
```javascript
for (let x in Object.keys({a: 1, b:2}) {
  doSomething(x);
}
```
It was in the middle of a section of code with some for-of loops and happened because I had just switched context from fixing a bug in a python client for this project.  Instead of calling `doSomething` with `a` then `b`, it instead called it with `0` then `1`.  This is expected behaviour for the for-in, but likely not what someone writing ES6 is looking to do.

##### Your take on the correct solution to problem.
I think that a non-default rule should be added to eslint to optionally warn/error on usage of the for-in construct.  I see this as similar to the existing `no-var` rule which exists to help people avoid footguns for those using ES6 features.  Using `var` in es6 code is just as valid as using `for (x in y)`.  I don't think this rule should be on by default, but it is general enough that having it in the linter itself would be a good choice.

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by nzakas** [comment]
I'm not sure saying to use for-of in place of for-in makes sense as a general rule. They do two entirely different things, as opposed to `var` and `let`, which do the same thing (declare a variable).

You might prefer for-of to forEach() or even the regular for loop, I'm just not sure it tracks to ban for-in. This might be better off as a custom rule.

**Comment by jhford** [comment]
> I'm not sure saying to use for-of in place of for-in makes sense as a general rule. They do two entirely different things, as opposed to var and let, which do the same thing (declare a variable).

var and let both declare a variable, but they do it completely differently.  Var does hoisting, can be redeclared multiple times and is function scope where let cannot be redeclared, does not hoist and is block scope.  Having a general rule to block makes as much sense as general rule as something to block using for-in.

I'm not trying to suggest that people ought to use for-of instead of for-in, just that there are cases where someone might want to prevent accidental use of for-in.

> You might prefer for-of to forEach() or even the regular for loop, I'm just not sure it tracks to ban for-in. This might be better off as a custom rule.

Because for-in in most other scripting languages is what for-of does in Javascript.  It's so easy to confuse, especially now that for-of is becoming more popular.  I don't think that this rule should be enabled by default, just like the no-var rule which is similar to this rule.  The point of this rule is to say that for-of replaces for-in, just that it's such a confusing naming situation that someone might want to be warned of its usage.

I already have this in my project as a custom rule, I just think that it's general enough and can see others using it.

**Comment by nzakas** [comment]
I'm sorry, I don't see this as general enough to include. We generally don't have rules that simply disallow syntax unless there's an obvious replacement or a known problem.
